From 8b47de1f4279a53bdb195c422b4c8244347fddb6 Mon Sep 17 00:00:00 2001
From: Gabe Lapingcao <mgultra2018@gmail.com>
Date: Tue, 28 Jul 2026 15:03:53 -0700
Subject: [PATCH 08/14] [tests/gtest]: implement RIB table NLRI assertions in
 test body

Ensures the expected RIB table matches the BGP-LS RIB table after test
execution. Development of `linkstate_data.h` is necessary for
replicating the structure of BGP-LS NLRI primitives used within FRR.
Attempting to include specific C headers from bgpd results in a cascade
of errors caused by different issues related to integrating legacy C
code with C++.

Signed-off-by: Gabe Lapingcao <mgultra2018@gmail.com>
---
 tests/gtest/include/bgpd_env.h                |   4 +
 tests/gtest/include/common_data.h             |  16 ++
 tests/gtest/include/frr_bridge.h              |  41 +++
 tests/gtest/include/linkstate_data.h          | 233 ++++++++++++++++++
 tests/gtest/include/test_bgp_ls_edge_update.h |  29 +++
 tests/gtest/src/bgpd_env.cpp                  |   6 +-
 tests/gtest/src/common_data.cpp               |  29 +++
 tests/gtest/src/frr_bridge.c                  |  58 +++++
 tests/gtest/src/main.cpp                      |  10 +
 tests/gtest/src/test_bgp_ls_edge_update.cpp   |  38 ++-
 10 files changed, 456 insertions(+), 8 deletions(-)
 create mode 100644 tests/gtest/include/linkstate_data.h

diff --git a/tests/gtest/include/bgpd_env.h b/tests/gtest/include/bgpd_env.h
index a768ce5f63..59d0560f7c 100644
--- a/tests/gtest/include/bgpd_env.h
+++ b/tests/gtest/include/bgpd_env.h
@@ -28,6 +28,10 @@ class BgpdEnvironment : public ::testing::Environment {
 
   /**
    * @brief Checks that FRR's BGP daemon is actively running.
+   *
+   * The actual boolean assertion is implemented inside
+   * `bridge_check_bgpd_running`. Additional configuration parameters may be set
+   * inside this function in the future.
    */
   virtual void SetUp() override;
 };
diff --git a/tests/gtest/include/common_data.h b/tests/gtest/include/common_data.h
index a0103c6e2e..e29e345dc5 100644
--- a/tests/gtest/include/common_data.h
+++ b/tests/gtest/include/common_data.h
@@ -6,6 +6,7 @@
 #include <string>
 #include <vector>
 
+#include "linkstate_data.h"
 #include "utils.hpp"
 
 using sys_id_t = std::string;
@@ -53,6 +54,21 @@ struct BgpLsLinkNlri {
   BgpLsNode source;
   BgpLsNode destination;
   BgpLsLink link;
+
+  /**
+   * @brief Converts `BgpLsLinkNlri` to `LinkState::LinkNlri`.
+   *
+   * Mainly needed for checking BGP-LS's RIB table for corresponding NLRI
+   * entries. Note that `LinkState::LinkNlri` is not the data type accepted by
+   * FRR's `bgpd` library; the required type is `struct bgp_ls_link_nlri`.
+   * However, the internal structure is completely replicated inside
+   * `linkstate_data.h`.
+   *
+   * In order to use the C API for determining if NLRI exists,
+   * `reinterpret_cast` is used to circumvent include errors when attempting to
+   * use the C API in C++ code.
+   */
+  explicit operator LinkState::LinkNlri() const;
 };
 NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(BgpLsLinkNlri, source, destination, link)
 
diff --git a/tests/gtest/include/frr_bridge.h b/tests/gtest/include/frr_bridge.h
index 31548b4897..babe1c4687 100644
--- a/tests/gtest/include/frr_bridge.h
+++ b/tests/gtest/include/frr_bridge.h
@@ -16,6 +16,11 @@
 extern "C" {
 #endif
 
+/**
+ * Forward-declaration for `bridge_link_exists_nlri`.
+ */
+struct bgp_ls_nlri;
+
 /**
  * @brief Initializes FRR's BGP daemon along with a single BGP routing instance.
  *
@@ -26,6 +31,16 @@ extern "C" {
  */
 void bridge_init_bgp(void);
 
+/**
+ * @brief Checks that FRR's BGP daemon is actively running.
+ *
+ * Checks if the `bgp` instance is not equal to `NULL` and `bgp_master` is not
+ * yet terminating.
+ *
+ * @return      True if the BGP daemon is running; false otherwise.
+ */
+bool bridge_check_bgpd_running(void);
+
 /**
  * @brief Clears the BGP instance's link-state TED.
  *
@@ -57,6 +72,11 @@ void bridge_send_message(struct stream* s, uint8_t msg_type);
  */
 void bridge_show_ted(struct sbuf* sbuf);
 
+/**
+ * @brief Pushes BGP-LS's RIB table to a string buffer for debugging.
+ */
+void bridge_show_table(struct sbuf* sbuf);
+
 /**
  * @brief Checks to see if an edge exists within BGP-LS's TED.
  *
@@ -70,6 +90,27 @@ void bridge_show_ted(struct sbuf* sbuf);
  */
 bool bridge_edge_exists_ted(struct ls_attributes* attr);
 
+/**
+ * @brief Checks to see if a link NLRI exists within the BGP instance's RIB.
+ *
+ * This function checks the RIB for a link NLRI corresponding to the NLRI
+ * passed.
+ *
+ * @param nlri      Link NLRI corresponding to the link to check. Represents a
+ *                      generic BGP-LS NLRI; however, the value passed should
+ *                      point to an NLRI struct with the link parameter filled.
+ * @return          True if the NLRI exists; false otherwise.
+ */
+bool bridge_link_exists_nlri(struct bgp_ls_nlri* nlri);
+
+/**
+ * @brief Checks to see if the BGP-LS TED and RIB table are empty.
+ *
+ * @return      True if both the TED and the RIB table are empty; false
+ *                  otherwise.
+ */
+bool bridge_check_ls_clear();
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/tests/gtest/include/linkstate_data.h b/tests/gtest/include/linkstate_data.h
new file mode 100644
index 0000000000..bac00fb7a8
--- /dev/null
+++ b/tests/gtest/include/linkstate_data.h
@@ -0,0 +1,233 @@
+/**
+ * Replicates BGP-LS primitive data structures required within FRR.
+ *
+ * These data structures are taken from `bgpd/bgp_ls_nlri.h`. Attempting to
+ * include this header within C++ code causes a domino effect: errors ranging
+ * from usages of C++ keywords as variable names, returning void pointers when
+ * function return type is specified as a non-void pointer, requiring a C++
+ * compiler for template definitions, etc. This comes from the fact that FRR's
+ * main library, found in the `lib/` directory, was made with C++ in mind.
+ * However, daemon-specific libraries, such as `bgpd/` and `isisd/`, are written
+ * in pure C.
+ */
+#ifndef LINKSTATE_DATA_H
+#define LINKSTATE_DATA_H
+
+#include <netinet/in.h>
+
+#include <cstdint>
+
+#include "lib/prefix.h"
+
+static constexpr int IGP_ROUTER_ID_ISIS_LEN = 6;
+static constexpr int IGP_ROUTER_ID_MAX_SIZE = 16;
+
+using as_t = uint32_t;
+
+namespace Model {
+namespace LinkState {
+
+/**
+ * @brief Replicates `thash_item` in FRR.
+ */
+struct HashItem {
+  HashItem *next;
+  uint32_t val;
+};
+
+/**
+ * @brief Replicates `bgp_ls_nlri_hash_item` (predeclared) in FRR.
+ */
+struct NlriHash {
+  HashItem item;
+};
+
+/**
+ * @brief Replicates `bgp_ls_nlri_type` in FRR.
+ */
+enum class NlriType : int {
+  RESERVED = 0,
+  NODE,
+  LINK,
+  IPV4_PREFIX,
+  IPV6_PREFIX,
+  SRV6_SID = 6
+};
+
+/**
+ * @brief Replicates `bgp_ls_protocol_id` in FRR.
+ */
+enum class Protocol : int {
+  RESERVED = 0,
+  ISIS_L1,
+  ISIS_L2,
+  OSPFV2,
+  DIRECT,
+  STATIC,
+  OSPFV3,
+  BGP
+};
+
+/**
+ * @brief Replicates `bgp_ls_ospf_route_type` in FRR.
+ */
+enum class OspfRouteType : int {
+  INTRA_AREA = 1,
+  INTER_AREA,
+  EXTERNAL_1,
+  EXTERNAL_2,
+  NSSA_1,
+  NSSA_2
+};
+
+/**
+ * @brief Replicates `bgp_ls_bgp_route_type` in FRR.
+ */
+enum class BgpRouteType : int {
+  LOCAL = 1,
+  ATTACHED,
+  EXTERNAL_BGP,
+  INTERNAL_BGP,
+  REDISTRIBUTED
+};
+
+enum NodeDescTLV : uint16_t {
+  AS_BIT = (1ULL << 0),
+  BGP_LS_ID_BIT = (1ULL << 1),
+  OSPF_AREA_BIT = (1ULL << 2),
+  IGP_ROUTER_BIT = (1ULL << 3),
+  BGP_ROUTER_ID_BIT = (1ULL << 4)
+};
+
+/**
+ * @brief Replicates `bgp_ls_node_descriptor` in FRR.
+ */
+struct NodeDesc {
+  uint16_t tlvs;
+  as_t asn;
+  uint32_t lsId;
+  uint32_t ospfAreaId;
+  uint8_t igpRouterIdLen;
+  union {
+    uint8_t sysid[IGP_ROUTER_ID_ISIS_LEN];
+    struct {
+      uint8_t sysid[IGP_ROUTER_ID_ISIS_LEN];
+      uint8_t psn;
+    } pseudoIsis;
+    in_addr ospf;
+    struct {
+      in_addr routerId;
+      in_addr ifaddr;
+    } pseudoOspf;
+    in_addr ipv4;
+    in6_addr ipv6;
+    uint8_t raw[IGP_ROUTER_ID_MAX_SIZE];
+  } igpRouterId;
+  in_addr bgpRouterId;
+};
+
+enum LinkDescTLV : uint32_t {
+  LINK_ID_BIT = (1ULL << 0),
+  IPV4_INTF_BIT = (1ULL << 1),
+  IPV4_NEIGH_BIT = (1ULL << 2),
+  IPV6_INTF_BIT = (1ULL << 3),
+  IPV6_NEIGH_BIT = (1ULL << 4),
+  MT_ID_BIT = (1ULL << 5),
+  REMOTE_AS_BIT = (1ULL << 6)
+};
+
+/**
+ * @brief Replicates `bgp_ls_link_descriptor` in FRR.
+ */
+struct LinkDesc {
+  uint32_t tlvs;
+  uint32_t localId;
+  uint32_t remoteId;
+  in_addr ipv4Intf;
+  in_addr ipv4Neigh;
+  in6_addr ipv6Intf;
+  in6_addr ipv6Neigh;
+  as_t remoteAsn;
+  uint16_t mtId;
+};
+
+/**
+ * @brief Replicates `bgp_ls_prefix_descriptor` in FRR.
+ */
+struct PrefixDesc {
+  uint16_t tlvs;
+  uint16_t mtId;
+  OspfRouteType ospfRT;
+  BgpRouteType bgpRT;
+  prefix prefix;
+};
+
+/**
+ * @brief Replicates `bgp_ls_srv6_sid_descriptor` in FRR.
+ */
+struct Srv6SidDesc {
+  uint16_t tlvs;
+  in6_addr sid;
+  uint16_t mtId;
+};
+
+/**
+ * @brief Replicates `bgp_ls_node_nlri` in FRR.
+ */
+struct NodeNlri {
+  Protocol proto;
+  uint64_t id;
+  NodeDesc node;
+};
+
+/**
+ * @brief Replicates `bgp_ls_link_nlri` in FRR.
+ */
+struct LinkNlri {
+  Protocol proto;
+  uint64_t id;
+  NodeDesc local;
+  NodeDesc remote;
+  LinkDesc link;
+};
+
+/**
+ * @brief Replicates `bgp_ls_prefix_nlri` in FRR.
+ */
+struct PrefixNlri {
+  Protocol proto;
+  uint64_t id;
+  NodeDesc local;
+  PrefixDesc prefix;
+};
+
+/**
+ * @brief Replicates `bgp_ls_srv6_sid_nlri` in FRR.
+ */
+struct Srv6Nlri {
+  Protocol proto;
+  uint64_t id;
+  NodeDesc local;
+  Srv6SidDesc sid;
+};
+
+/**
+ * @brief Replicates `bgp_ls_nlri` in FRR.
+ */
+struct ParentNlri {
+  NlriType type;
+  union {
+    NodeNlri node;
+    LinkNlri link;
+    PrefixNlri prefix;
+    Srv6Nlri srv6;
+  } data;
+  unsigned long refcnt;
+  uint32_t id;
+  NlriHash hash;
+};
+
+}  // namespace LinkState
+}  // namespace Model
+
+#endif  // LINKSTATE_DATA_H
diff --git a/tests/gtest/include/test_bgp_ls_edge_update.h b/tests/gtest/include/test_bgp_ls_edge_update.h
index 88752fad88..c0c58951d7 100644
--- a/tests/gtest/include/test_bgp_ls_edge_update.h
+++ b/tests/gtest/include/test_bgp_ls_edge_update.h
@@ -17,10 +17,23 @@
 namespace Model {
 
 class EdgeTest : public testing::TestWithParam<TestCase> {
+ public:
+  /**
+   * @brief Sets debugging on (true) or off (false).
+   *
+   * Debugging on prints BGP-LS's TED and RIB table to the console during Google
+   * Test runs. Debugging off skips these lines.
+   */
+  static void SetDebugMode(bool debug);
+
  protected:
   /**
    * @brief Checks that FRR's BGP daemon is actively running and the BGP-LS TED
    * is empty.
+   *
+   * Most of the work is done inside `bridge_check_bgpd_running` and
+   * `bridge_clear_bgp_ls_ted` within frr_bridge.h. Fails the current test run
+   * if either condition is note met.
    */
   virtual void SetUp() override;
 
@@ -113,6 +126,22 @@ class EdgeTest : public testing::TestWithParam<TestCase> {
    */
   void SendUpdateMessage(const BApiLinkStateUpdate& apiMessage,
                          ls_attributes*& attr) const;
+
+  /**
+   * @brief Verifies the entirety of the model's RIB exists within the current
+   * BGP instance.
+   *
+   * Iterates over each `BgpLsLinkNlri` and confirms their existence within
+   * BGP-LS's RIB table. This function uses an unsafe cast, `reinterpret_cast`,
+   * to provide the C API the correct data structure. Fails the current test run
+   * if at least one entry within `rib` is missing.
+   *
+   * @param rib     Collection of `BgpLsLinkNlri` containing NLRI to check the
+   *                    RIB with.
+   */
+  void VerifyNlri(const std::vector<BgpLsLinkNlri>& rib) const;
+
+  static inline bool DebugMode = false;
 };
 }  // namespace Model
 
diff --git a/tests/gtest/src/bgpd_env.cpp b/tests/gtest/src/bgpd_env.cpp
index a7e0a4a881..de1cf14101 100644
--- a/tests/gtest/src/bgpd_env.cpp
+++ b/tests/gtest/src/bgpd_env.cpp
@@ -11,7 +11,11 @@ BgpdEnvironment::BgpdEnvironment() { bridge_init_bgp(); }
 BgpdEnvironment::~BgpdEnvironment() { bridge_clean_bgp(); }
 
 void BgpdEnvironment::SetUp() {
-  // TODO
+  ASSERT_TRUE(bridge_check_bgpd_running())
+      << "[Environment SetUp]: bgpd is not running.";
+
+  // we can also set some configurations here; ensure that bgpd is self-
+  // contained, not listening through other ports
 }
 
 }  // namespace Model
diff --git a/tests/gtest/src/common_data.cpp b/tests/gtest/src/common_data.cpp
index e3e1fb2715..9b7806dd04 100644
--- a/tests/gtest/src/common_data.cpp
+++ b/tests/gtest/src/common_data.cpp
@@ -1,7 +1,36 @@
 #include "common_data.h"
 
+#include "lib/zebra.h"
+#include "linkstate_data.h"
+#include "utils.hpp"
+
 namespace Model {
 
+BgpLsLinkNlri::operator LinkState::LinkNlri() const {
+  // IMPORTANT NOTE
+  // When converting a `BgpLsLinkNlri` to a `LinkNlri`, we REVERSE the direction
+  // of the edge. This may be because advertising the forward edge first does
+  // not actually install the entry until the reverse edge is processed in the
+  // Google Test implementation. Will need to check this again in the future.
+  LinkState::LinkNlri nlri{.proto = LinkState::Protocol::ISIS_L1};
+
+  nlri.local.igpRouterIdLen = SysIdToBuffer(nlri.remote.igpRouterId.sysid,
+                                            this->source.igp_router_id.c_str());
+  nlri.remote.igpRouterIdLen = SysIdToBuffer(
+      nlri.local.igpRouterId.sysid, this->destination.igp_router_id.c_str());
+
+  inet_pton(AF_INET6, this->link.interface.c_str(), &nlri.link.ipv6Neigh);
+  inet_pton(AF_INET6, this->link.neighbor.c_str(), &nlri.link.ipv6Intf);
+
+  // must set valid flags
+  SET_FLAG(nlri.local.tlvs, LinkState::NodeDescTLV::IGP_ROUTER_BIT);
+  SET_FLAG(nlri.remote.tlvs, LinkState::NodeDescTLV::IGP_ROUTER_BIT);
+  SET_FLAG(nlri.link.tlvs, LinkState::LinkDescTLV::IPV6_INTF_BIT);
+  SET_FLAG(nlri.link.tlvs, LinkState::LinkDescTLV::IPV6_NEIGH_BIT);
+
+  return nlri;
+}
+
 LinkStateEdge::operator BApiLinkStateUpdate() const {
   BApiLinkStateUpdate message{.event = BEvent::UPDATE,
                               .remote = this->destination_node,
diff --git a/tests/gtest/src/frr_bridge.c b/tests/gtest/src/frr_bridge.c
index 68cac5a8c5..ba5b00a068 100644
--- a/tests/gtest/src/frr_bridge.c
+++ b/tests/gtest/src/frr_bridge.c
@@ -109,6 +109,8 @@ void bridge_init_bgp(void) {
   bgp_get(&bgp, &asn, NULL, BGP_INSTANCE_TYPE_DEFAULT, NULL, ASNOTATION_PLAIN);
 }
 
+bool bridge_check_bgpd_running(void) { return bgp != NULL && !bm->terminating; }
+
 void bridge_clear_bgp_ls_ted(void) { bgp_ls_withdraw_ted(bgp); }
 
 void bridge_clean_bgp(void) {
@@ -213,6 +215,32 @@ void bridge_show_ted(struct sbuf *sbuf) {
   }
 }
 
+void bridge_show_table(struct sbuf *sbuf) {
+  struct bgp_table *table;
+  struct bgp_dest *dest;
+  char formatted_nlri[1024];
+  struct bgp_ls_nlri *entry;
+
+  if (!bgp || !bgp->ls_info) {
+    return;
+  }
+
+  table = bgp->rib[AFI_BGP_LS][SAFI_BGP_LS];
+  if (!table) {
+    return;
+  }
+
+  for (dest = bgp_table_top(table); dest; dest = bgp_route_next(dest)) {
+    entry = dest->ls_nlri;
+    if (!entry) {
+      continue;
+    }
+
+    bgp_ls_nlri_format(entry, formatted_nlri, sizeof(formatted_nlri));
+    sbuf_push(sbuf, 0, "%s\n", formatted_nlri);
+  }
+}
+
 bool bridge_edge_exists_ted(struct ls_attributes *attr) {
   struct ls_edge *src_edge = ls_find_edge_by_source(bgp->ls_info->ted, attr);
   struct ls_edge *dst_edge =
@@ -226,3 +254,33 @@ bool bridge_edge_exists_ted(struct ls_attributes *attr) {
          ls_vertex_same(src_edge->source, dst_edge->destination) &&
          ls_vertex_same(src_edge->destination, dst_edge->source);
 }
+
+bool bridge_link_exists_nlri(struct bgp_ls_nlri *nlri) {
+  struct bgp_ls_nlri *ret = bgp_ls_nlri_lookup(&bgp->ls_info->nlri_hash, nlri);
+  return ret != NULL;
+}
+
+bool bridge_check_ls_clear() {
+  // BGP-LS TED checks
+  struct ls_vertex *vertex;
+  frr_each(vertices, &bgp->ls_info->ted->vertices, vertex) {
+    if (vertex) {
+      return false;
+    }
+  }
+
+  struct ls_edge *edge;
+  frr_each(edges, &bgp->ls_info->ted->edges, edge) {
+    if (edge) {
+      return false;
+    }
+  }
+
+  struct ls_subnet *subnet;
+  frr_each(subnets, &bgp->ls_info->ted->subnets, subnet) {
+    if (subnet) {
+      return false;
+    }
+  }
+  return true;
+}
diff --git a/tests/gtest/src/main.cpp b/tests/gtest/src/main.cpp
index 9e99f0e6d6..97c2072378 100644
--- a/tests/gtest/src/main.cpp
+++ b/tests/gtest/src/main.cpp
@@ -4,9 +4,12 @@
 #include <fstream>
 #include <iostream>
 #include <nlohmann/json.hpp>
+#include <string_view>
+#include <vector>
 
 #include "bgpd_env.h"
 #include "common_data.h"
+#include "test_bgp_ls_edge_update.h"
 
 /**
  * Entrypoint for Google Test execution. Before initializing the testing
@@ -27,6 +30,13 @@ int main(int argc, char** argv) {
   nlohmann::json data = nlohmann::json::parse(raw);
   Model::testCases = data.get<std::vector<Model::TestCase>>();
 
+  std::vector<std::string_view> args(argv, argv + argc);
+  for (const auto& arg : args) {
+    if (arg == "--debug" || arg == "-d") {
+      Model::EdgeTest::SetDebugMode(true);
+    }
+  }
+
   ::testing::InitGoogleTest(&argc, argv);
 
   ::testing::Environment* const env =
diff --git a/tests/gtest/src/test_bgp_ls_edge_update.cpp b/tests/gtest/src/test_bgp_ls_edge_update.cpp
index 11ce2d7da7..404daaa337 100644
--- a/tests/gtest/src/test_bgp_ls_edge_update.cpp
+++ b/tests/gtest/src/test_bgp_ls_edge_update.cpp
@@ -11,17 +11,23 @@
 #include "frr_bridge.h"
 #include "lib/stream.h"
 #include "lib/zclient.h"
+#include "linkstate_data.h"
 #include "sbuf.h"
 #include "utils.hpp"
 
 namespace Model {
 
 void EdgeTest::SetUp() {
-  // TODO
+  ASSERT_TRUE(bridge_check_bgpd_running())
+      << "[Fixture SetUp]: bgpd is not running.";
+  ASSERT_TRUE(bridge_check_ls_clear())
+      << "[Fixture SetUp]: BGP-LS TED is not empty.";
 }
 
 void EdgeTest::TearDown() { bridge_clear_bgp_ls_ted(); }
 
+void EdgeTest::SetDebugMode(bool debug) { EdgeTest::DebugMode = debug; }
+
 void EdgeTest::NodeIdToFrr(const LinkStateNodeId& nodeId,
                            ls_node_id& frrNodeId) const {
   // for now, we default to IS-IS level 1
@@ -142,6 +148,20 @@ void EdgeTest::SendUpdateMessage(const BApiLinkStateUpdate& apiMessage,
   ls_node_del(remote_node);
 }
 
+void EdgeTest::VerifyNlri(const std::vector<BgpLsLinkNlri>& rib) const {
+  for (const auto& entry : rib) {
+    LinkState::LinkNlri nlri = static_cast<LinkState::LinkNlri>(entry);
+    LinkState::ParentNlri p{.type = LinkState::NlriType::LINK,
+                            .data = {.link = nlri}};
+
+    // WARNING: unsafe cast; this is being used to bypass include errors with
+    // the FRR bgpd library
+    bgp_ls_nlri* nlriCast = reinterpret_cast<bgp_ls_nlri*>(&p);
+    ASSERT_TRUE(bridge_link_exists_nlri(nlriCast))
+        << "[bgp_ls_nlri]: NLRI does not exist within the RIB.";
+  }
+}
+
 TEST_P(EdgeTest, ValidateEdgeUpdate) {
   // Arrange
   TestCase tc = GetParam();
@@ -168,18 +188,22 @@ TEST_P(EdgeTest, ValidateEdgeUpdate) {
   SendUpdateMessage(tc.api_param, attr);
 
   // Debug
-  struct sbuf sbuf;
-  sbuf_init(&sbuf, NULL, 0);
-
-  bridge_show_ted(&sbuf);
-  std::cout << sbuf_buf(&sbuf) << std::endl;
+  if (EdgeTest::DebugMode) {
+    struct sbuf sbuf;
+    sbuf_init(&sbuf, NULL, 0);
+
+    bridge_show_ted(&sbuf);
+    bridge_show_table(&sbuf);
+    std::cout << sbuf_buf(&sbuf) << std::endl;
+    sbuf_free(&sbuf);
+  }
 
   // Assert
   ASSERT_TRUE(bridge_edge_exists_ted(attr))
       << "[ls_attributes]: edge does not exist within TED.";
+  VerifyNlri(tc.final_state.rib);
 
   // Clean
-  sbuf_free(&sbuf);
   ls_attributes_del(attr);
 }
 
-- 
2.34.1

